ATR-759: Add release notes for 3.1.2, update 1007 diagnostic description (#496)
Acumatica/Acuminator: docs/ReleaseNotes.md

# Acuminator Release Notes
This document provides information about fixes, enhancements, and key features that are available in Acuminator.

## Acuminator 3.1.2 - 2023-06-06
Acuminator 3.1.2 includes the bug fixes and enhancements described in this section, as well as the features that have been implemented in previous versions.

### Enhancements
In Acuminator 3.1.2, the following enhancements have been implemented:

>  - The [PX1096](diagnostics/PX1096.md) diagnostic has been added to check signature of methods with the `PXOverride` attribute: The diagnostic checks that the base graph or base graph extensions contains a base method that is compatible with the method with the `PXOverride` attribute. 

Reviewer: Maybe rephrase it like this?

```suggestion
 - The [PX1096](diagnostics/PX1096.md) diagnostic has been added to check signature of methods with the `PXOverride` attribute: The diagnostic checks whether the base graph or base graph extensions contain a base method  compatible with the method with the `PXOverride` attribute. 
```

 - The [PX1007](diagnostics/PX1007.md) diagnostic no longer checks the following DAC fields for a code annotation:
   - `Selected`
   - `Attributes`
   - `GroupMask`
   - `CompanyID`
   - `CompanyMask`
   - `DeletedDatabaseRecord`
 - The [PX1007](diagnostics/PX1007.md) diagnostic has been updated in the following ways:

>    - The diagnostic now supports the `inheritdoc` tag for DACs, DAC extensions, and their properties. The diagnostic also checks that mapped DAC field properties have the `inheritdoc` tag with the `cref` attribute pointing to the mapped DAC field property of the original DAC. 

Reviewer: ```suggestion
   - The diagnostic now supports the `inheritdoc` tag for DACs, DAC extensions, and their properties. The diagnostic also checks that DAC field properties of projection DACs have the `inheritdoc` tag with the `cref` attribute pointing to the mapped DAC field property of the original DAC. 
```

>    - A new code fix can generate the `inheritdoc` tag for a mapped DAC property of a projection DAC.

Reviewer: It's worth mentioning that VS allows us to apply code fix quickly to multiple alerts with its "Fix All" functionality for code fix.

Author: I have mentioned it in the diagnosic description., so added just a small mention here

   - Code fixes of the diagnostic can now generate documentation correctly even for a badly formatted code and remove incorrect documentation tags. The code fixes also support doc tags declared as empty XML elements such as `<summary/>`.

>    - The default justification "to be documented later" has been removed from the diagnostic. To remove the DAC or its property from the documentation, use the `exclude` tag.

Reviewer: ```suggestion
   - The default justification "to be documented later" has been removed from the diagnostic. To remove a DAC, DAC extension or DAC field property from the documentation, use the `exclude` tag.
```

# Acuminator Release Notes
This document provides information about fixes, enhancements, and key features that are available in Acuminator.

> ## Acuminator 3.1.2

Reviewer: It would be good to start adding dates to releases

Acuminator 3.1.2 includes the bug fixes and enhancements described in this section, as well as the features that have been implemented in previous versions.

### Enhancements
In Acuminator 3.1.2, the following enhancements have been implemented:

>  - The [PX1096](diagnostics/PX1096.md) diagnostic has been added to check signature of methods with the `PXOverride` attribute.

Reviewer: I would rephrase this a bit, it's not just a some kind of generic unspecified check of PXOverrides. It's a check that in the base graph and base graph extensions there exists a base method that is compatible with the PXOverride method. This means that the base method has the same name, its accessibility public, protected or protected internal, and it has a compatible parameters.  More about this here:
- https://help.acumatica.com/(W(147))/Help?ScreenId=ShowWiki&pageid=6fa2a444-17b4-42f9-9e6a-64e85167626a
- https://help.acumatica.com/(W(146))/Help?ScreenId=ShowWiki&pageid=635c830e-4617-4d5c-9fa5-035952311aa9

>  - The [PX1007](diagnostics/PX1007.md) diagnostic no longer checks the following DAC fields for a code annotation:

Reviewer: The default justification for PX1007 was deleted, this should be mentioned.

   - `Selected`
   - `Attributes`
   - `GroupMask`
   - `CompanyID`
   - `CompanyMask`
   - `DeletedDatabaseRecord`

>  - The [PX1007](diagnostics/PX1007.md) diagnostic now checks that mapped DAC fields of projection DACs have the `inheritoc` XML tag

Reviewer: That's not complete description + there is a mistype in the tag name. 
PX1007 now checks that mapped DAC field **properties** have `inheritdoc` tag with `cref` attribute pointing to the mapped DAC field property of the original DAC. An example of correctly declared projection DAC property will be really helpful here and a link to the inheritdoc tag documentation:
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/recommended-tags#inheritdoc

Author: Is it okay if I give the link to inheritdoc documnetation and details about the original mapped property in the article about PX1007? This is too specific info for RNs

Reviewer: Yes, it's okay.

>  

Reviewer: The PX1007 changes description should also include that PX1007 supports  `inheritdoc` tag in general, not just for projection DAC mapped properties. Basically, you can put any form of inheritdoc tag on normal DACs, DAC extensions and their properties but you have to follow particular rule for projection DAC mapped DAC properties.

>  

Reviewer: There is also no description for a new code fix which will generate correct inheritdoc tag for a mapped DAC property of a projection DAC. It's worth mentioning that VS allows us to apply single code fix to multiple alerts with its "Fix All" functionality for code fixes. It means that this code fix can be used to very quickly automatically generate correct documentation for most of the projection DAC properties. Before we had to do it manually which is a rather tedious task.

Reviewer: this remark has not been covered yet

>  

Reviewer: Another change that is worth mentioning is that now all PX1007 code fixes generate documentation correctly even for a badly formatted code, and will keep existing documentation tags (except the tags deleted by the inheritdoc tag generation code fix I described in the previous comment)

Also the diagnostic and code fixes now support doc tags declared as empty XML elements like `<summary/>`

You can use the PR overview for the list of al PX1007 changes https://github.com/Acumatica/Acuminator/pull/497

Author: Could you plesae clarify what is a badly formatted code?

Reviewer: For example, the code without the empty lines between elements:
```csharp
#region OrderType
public abstract class orderType : IBqlField { }
[PXDBString(IsKey = true, InputMask = "")]
[PXDefault]
[PXUIField(DisplayName = "Order Type")]
public string OrderType { get; set; }
#endregion
```

### Fixed Bugs
In this version of Acuminator, the following bugs have been fixed:
 - The [PX1023](diagnostics/PX1023.md) diagnostic showed false alert for the `INUnit` attribute.

>  - In Visual Studio 2022, Acuminator analysis did not work in the out-of-process mode (the **Run code analysis in separate process** check box in Visual Studio options)

Reviewer: Maybe we should emphasize this bug a little? Maybe place it first? It's actually a blocker that was a reason for an early release.

>  - It was possible to suppress the [PX1007](diagnostics/PX1007.md) diagnostic for a class with the `PXCacheExtension` attribute by marking the class with the `PXHidden` attribute

Reviewer: This is not correct. `PXCacheExtension` is not an attribute, its a base type for DAC extensions. Hence sometimes DAC extensions are also called cache extensions. The more correct version would be something like:
```
 It was possible to suppress the [PX1007](diagnostics/PX1007.md) diagnostic for DAC extensions by marking them with the `PXHidden` attribute
```

 - The [PX1016](diagnostics/PX1016.md) diagnostic showed a false alert for generic non-abstract graph extensions.

## Gratitudes 

> The Acuminator team would like to thank our open source contributor @taskinkemal. We are grateful for your input and hope you stay with us.

Reviewer: Imho, we should mention a direct name together with the github account, it's Kemal Taskin.
